docs: Devcontainers-cli enhancement (#2548)
zmkfirmware/zmk: docs/docs/development/local-toolchain/build-flash.mdx

    them in the form of `-DSHIELD=kyra_left`.

    :::tip
    Once the first generation of the build directory using one-time CMake
    arguments is done, you can execute subsequent builds by using:

    ```sh
    west build -d <build_dir>
    ```
    :::

  </TabItem>
  <TabItem value="permanent">
    The `west build` command can be configured using
    [`west config`](https://docs.zephyrproject.org/3.5.0/develop/west/config.html#west-config-cmd).
    This makes it possible to add permanent CMake arguments, which are invoked
    whenever a new build system is generated. To do this, we make use of its
    `build.cmake-args` configuration option.

    ```sh
    west config build.cmake-args -- -DSHIELD=kyra_left
    ```

    For adding multiple arguments a string can be used:

    ```sh
    west config build.cmake-args \
      -- "-DSHIELD=kyra_left -DZMK_CONFIG=/absolute/path/to/zmk-config"
    ```

    :::note
    New build systems are only generated with `west build` when they are
    currently is non-available. Therefore, either delete the build directory or
    do a **pristine build**.
    :::

  </TabItem>

> </Tabs>

Author: I removed the tabs and did some small adjustments to the text, was this what you had in mind?

with an onboard MCU or one that uses an MCU board add-on.

<Tabs defaultValue="onboardMcu"
      values={[
        {label: 'Onboard MCU', value: 'onboardMcu'},
        {label: 'Addon MCU', value: 'addonMcu'}
      ]}>
  <TabItem value="onboardMcu">
    Keyboards with onboard MCU chips are simply treated as the
    [board](https://docs.zephyrproject.org/3.5.0/hardware/porting/board_porting.html)
    as far as Zephyr™ is concerned.

    Given the following:

    - Keyboard: Planck (rev6)
    - Keymap: default

    you can build ZMK with the following:

    ```sh
    west build -b planck_rev6
    ```

  </TabItem>
  <TabItem value="addonMcu">
    ZMK treats keyboards that take an MCU addon board as
    [shields](https://docs.zephyrproject.org/3.5.0/hardware/porting/shields.html),
    and treats the smaller MCU board as the true
    [board](https://docs.zephyrproject.org/3.5.0/hardware/porting/board_porting.html).

    Given the following:

    - MCU Board: Proton-C
    - Keyboard PCB: kyria_left
    - Keymap: default

    You can build ZMK with the following:

    ```sh
    west build -b proton_c -- -DSHIELD=kyria_left
    ```

  </TabItem>
</Tabs>

### CMake Arguments

> For creating a build system, the `west build` command uses
> [CMake](https://cmake.org/). This allows for additional arguments to be added to
> its invocations.

Reviewer: ```suggestion
The `west build` command builds devices using [CMake](https://cmake.org/). In addition to `west build`'s configuration options, arguments can be passed to CMake directly.
```

> The simplest of doing this is by adding `--` after the `west build` command and
> listing the CMake arguments afterward. For example, the above-shown command for
> building a keyboard with an MCU board add-on uses them in the form of
> `-DSHIELD=kyra_left`.

Reviewer: ````suggestion
This is done by adding `--` after the `west build` command and
listing the CMake arguments afterward. The previous section shows one example of this with `-DSHIELD=kyria_left`. Another example use case is passing Kconfig flags:
```sh
west build -b planck_rev6 -- -DCONFIG_ZMK_SLEEP=y
```
````

:::tip

> Once the first generation of the build directory using one-time CMake
> arguments is done, you can execute subsequent builds by using:

Reviewer: ```suggestion
Once the first generation of the build directory using one-time CMake
arguments is done, you can omit the CMake arguments and board selection, instead building with the command:
```

```sh
west build -d <build_dir>
```

:::

> Furthermore, it is possible to configure `west build` by using
> [`west config`](https://docs.zephyrproject.org/3.5.0/develop/west/config.html#west-config-cmd).
> This makes it possible to add permanent CMake arguments, which are invoked
> whenever a new build system is generated. To do this, we make use of its
> `build.cmake-args` configuration option.

Reviewer: ```suggestion
You can also add permanent CMake arguments to `west build` by using the
[`west config`](https://docs.zephyrproject.org/3.5.0/develop/west/config.html#west-config-cmd) command. These are invoked whenever a new build system is generated. To add permanent arguments, set the `build.cmake-args` configuration option.
```

```sh
west config build.cmake-args -- -DSHIELD=kyra_left
```

> For adding multiple arguments a string can be used:

Reviewer: ```suggestion
Multiple arguments are added by assigning a string to the configuration option:
```

```sh
west config build.cmake-args \
  -- "-DSHIELD=kyra_left -DZMK_CONFIG=/absolute/path/to/zmk-config"
```

> :::note
> New build systems are only generated with `west build` when they are
> currently is non-available. Therefore, either delete the build directory or
> do a **pristine build**.
> :::
> 

Reviewer: ```suggestion
```
I don't think this note is necessary as pristine building is discussed immediately afterwards.

## Building

To build for your particular keyboard, the behavior varies slightly depending on
if you are building for a keyboard with an onboard MCU, or one that uses an MCU
board addon.

:::warning
From here on the process for building and flashing varies based on the method
you're using:

- **[VS Code](setup/docker/vscode.md)/[Native](setup/native.mdx)**:
  Execute the commands from within the `app/` subdirectory of the ZMK checkout.
- **[Devcontainer CLI](setup/docker/cli.md)**:
  Add `-s /absolute/path/to/zmk/app` to the `west` command.

If this is not done, you will encounter errors such as: `ERROR: source directory
"." does not contain a CMakeLists.txt; is this really what you want to build?`

> :::

Author: I updated everything according to my last comment. Hope this resolves this issue.

### Building from `zmk-config` Folder

> 

Author: I do agree adding documentation for [Permanent CMake Arguments](https://docs.zephyrproject.org/3.5.0/develop/west/build-flash-debug.html#id17) is most likely interesting for a lot of people. But I do think adding it at the marked location is not a good idea since most of the commands use such arguments. I think adding it as a general information at the top of the file is better. What do you think?

I do have to say I do not really understand what is mentioned in second link regarding the [west config](https://docs.zephyrproject.org/3.5.0/develop/west/config.html#west-config-cmd). Guess that is due to the fact that I am not that familiar with `west`. Would you mind explaining that to me?

> 

Reviewer: The west config note is just a minor thing. If we mention permanent cmake arguments then we need to mention west config, as the permanent arguments are set using west config.

> 

Author: I did my best to write something appropriate. If anything is wrong or needs adjustment let me know.

### Building from `zmk-config` Folder

Instead of building .uf2 files using the default keymap and config files, you

> can build directly from your [`zmk-config` folder](../../user-setup.mdx#github-repo)

Reviewer 2: ```suggestion
can build using files from your [`zmk-config` folder](../../user-setup.mdx#github-repo)
```
This seems a bit wrong to me, since you aren't building while inside the config folder.

(I am aware this isn't coming from the new commits, but if we are touching here we can improve.)

by adding `-DZMK_CONFIG="C:/the/absolute/path/config"` to your `west build`

> command. **Notice that this path should point to the folder labelled `config`

Reviewer 2: ```suggestion
command. **Notice that this path should point to the folder labeled `config`
```
